src/Simulation/Simulators/OpenSystemsSimulator/DataModel/ArrayConverter.cs

```diff
@@ -3,249 +3,245 @@
-                        data = new List<Complex>();
-                        reader.Require(JsonTokenType.StartArray);
-
-                        while (reader.Read() && reader.TokenType != JsonTokenType.EndArray)
-                        {
-                            // We expect an inner array at this point, so
-                            // go on and read the begin array token.
-                            reader.Require(JsonTokenType.StartArray, read: false);
-
-                            reader.Read();
-                            var real = reader.GetDouble();
-
-                            reader.Read();
-                            var imag = reader.GetDouble();
-
-                            reader.Require(JsonTokenType.EndArray);
-
-                            data.Add(new Complex(real, imag));
-                        }
-                        break;
-
-                    default:
-                        throw new JsonException();
-                }
-            }
+                        var real = reader.GetDouble();
 
-            // At this point, none of version, dims, or data should be null.
-            if (version == null) throw new JsonException(nameof(version));
-            if (dims == null) throw new JsonException(nameof(dims));
-            if (data == null) throw new JsonException(nameof(data));
-
-            // We now know we have everything we need to make the new array.
-            // In doing so, we allocate an ndarray with of shape (nElements, 2)
-            // where the last index represents real-vs-imag. We'll reshape
-            // it to the correct shape at the end.
-            var array = np.zeros((data.Count, 2));
-            foreach (var idx in Enumerable.Range(0, data.Count))
-            {
-                var element = data[idx];
-                array[idx, 0] = element.Real;
-                array[idx, 1] = element.Imaginary;
+                        reader.Read();
+                        var imag = reader.GetDouble();
+
+                        reader.Require(JsonTokenType.EndArray);
+
+                        data.Add(new Complex(real, imag));
+                    }
+                    break;
+
+                default:
+                    throw new JsonException();
             }
-            return np.reshape(array, dims.Concat(new [] { 2 }).ToArray());
         }
 
-        public override void Write(Utf8JsonWriter writer, NDArray value, JsonSerializerOptions options)
+        // At this point, none of version, dims, or data should be null.
+        if (version == null) throw new JsonException(nameof(version));
+        if (dims == null) throw new JsonException(nameof(dims));
+        if (data == null) throw new JsonException(nameof(data));
+
+        // We now know we have everything we need to make the new array.
+        // In doing so, we allocate an ndarray with of shape (nElements, 2)
+        // where the last index represents real-vs-imag. We'll reshape
```

Comment:
a. The wording "an ndarray **with of** shape" seems odd to me.
b. FYI: I don't understand the fragment "the last index represents real-vs-imag".

Reply:
(a) was a typo, thanks for catching that. (b) indicates that the last index is of length two, and indicates whether each element describes the real or imaginary part of a complex number. For example, the 2D array `[[1, i], [-i, 1]]` would get represented by first flattening to `[1, i, -i, 1]` (where `dims` indicates how to unflatten again), then expanding out complex numbers, `[[1, 0], [0, 1], [0, -1], [1, 0]]`.

Reply:
Expanded comment.